Add broken pair reader

betatim opened the pull request:
WIP

Copy functionality of `utils.broken_paired_reader` to C++

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [ ] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [ ] Did it change the command-line interface? Only backwards-compatible
  additions are allowed without a major version increment. Changing file
  formats also requires a major version number increment.
- [ ] For substantial changes or changes to the command-line interface, is it
  documented in `CHANGELOG.md`? See [keepachangelog](http://keepachangelog.com/)
  for more details.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)

betatim commented:
Totally not optimised c++, benchmarking `time scripts/extract-paired-reads.py ecoli_ref-5m.fastq.gz -s /dev/null -p /dev/null`:

* `master`: 71.77s user 0.53s system 99% cpu 1:12.65 total
* this PR: 56.67s user 0.89s system 98% cpu 58.397 total

😕

ctb commented:
Do you think that's because the Python side is slowing stuff down a lot?

betatim commented:
There are a few things that I am confused about/need some educating.

Why the `PyObject * obj = khmer_ReadPairIterator_Type.tp_alloc(&khmer_ReadPairIterator_Type, 1);` construct instead of `PyObject_Init` and friends? It seems the segfault (on gcc4.9 only) in the previous commit was due to having a `Read` member in `khmer_ReadPairIterator_Object` instead of a `Read *`.

This version also runs in about ~53s. Next up investigating why this doesn't run faster and maybe a small toy example to fully understand how memory gets initialised for python instances :-/ My guess as to why this isn't faster would be that we switch from C to python to C to python. Instead of handing a big chunk of work off to C land (`numpy` style).

ctb commented:
Compare with @camillescott's Cythonized `broken_paired_reader`:

```
In [5]: %timeit [record for record in BrokenPairedReader(FastxParser('test-mixed.fa'))]
100 loops, best of 3: 3.49 ms per loop

In [6]: %timeit [record for record in broken_paired_reader(screed.open('test-mixed.fa'))]
100 loops, best of 3: 17.8 ms per loop
```

betatim commented:
Exactly, it would suggest that there is something obviously wrong with my c++ 😢

betatim commented:
With dc0d0a6, modified `setup.py` (to get it compile), and modified `extract-paired-reads.py` like so:
```python
from khmer._oxli import streaming
screed_iter = streaming.FastxParser(infile)
for index, is_pair, read1, read2 in streaming.BrokenPairedReader(screed_iter):
    ...
```
the above bench on ecoli takes ~60s. Conclusion: we can make the broken paired read parsing faster (yay!) but that isn't where we are spending our time (at least in this application) (:cry:)

ctb commented:
what is your thinking on this now?

betatim commented:
Still think that reading stuff from disk doesn't seem to be the bottle neck. However both Camille's cython and this PR are quicker than the python version, so let's merge it.

betatim commented:
New machine, new benchmarks (`time scripts/extract-paired-reads.py ecoli_ref-5m.fastq.gz -s /dev/null -p /dev/null`).

* `master`: 31s
* this branch: 23s

This is pretty big! I am worried about correctness but all the tests pass 😕 The code is quite messy as it is, I will tidy it up a bit and think through that gain more confidence in it 🤞.

ctb commented:
sounds good :) - thx!

betatim commented on scripts/extract-paired-reads.py:
```diff
@@ -154,8 +153,12 @@ def main():
+    # python version
+    # for index, is_pair, read1, read2 in broken_paired_reader(reads_iter):
+    # c++ version
+    for index, (read1, read2) in enumerate(reads_iter.iter_read_pairs(0)):
```
Should be able to use a symbol for this.

betatim commented:
I thought it would be a good idea to get this done before #1591 but somehow nearly ended in disaster. Tried to replace the internals of `broken_paired_reader` so that they can take reads from `ReadParser(fname).iter_...` and `screed.open(fname)`. New attempt.

betatim commented:
Thoughts on punting this till after 2.0? The code proposed in this PR is pretty messy and ergh, post 2.0 there is the ope of cython where this work has already been done. So I would close this PR and reevaluate after 2.0

ctb commented:
+1

On Fri, Apr 07, 2017 at 02:23:05AM -0700, Tim Head wrote:
> Thoughts on punting this till after 2.0? The code proposed in this PR is pretty messy and ergh, post 2.0 there is the ope of cython where this work has already been done. So I would close this PR and reevaluate after 2.0
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/khmer/pull/1538#issuecomment-292487016
-- 
C. Titus Brown, ctbrown@ucdavis.edu

betatim commented:
Closed in favour of cython (or a new attempt).

---

Please don't delete the branch so we can recover the code from this PR.
